pilot/pkg/networking/core/v1alpha3/envoyfilter.go

```diff
@@ -356,30 +443,253 @@ func applyListenerConfigPatches(listeners []*xdsapi.Listener, env *model.Environ
+			continue
+		}
+
+		if cp.Patch.Operation != networking.EnvoyFilter_Patch_REMOVE {
+			continue
+		}
+
+		// iterate through all virtual hosts in a route and remove ones that match
+		for i := range routeConfiguration.VirtualHosts {
+			if routeConfigurationMatch(routeConfiguration, &routeConfiguration.VirtualHosts[i], pluginParams, cp.Match) {
+				// set name to empty. We remove virtual hosts with empty names later in this function
+				routeConfiguration.VirtualHosts[i].Name = ""
+				virtualHostsRemoved = true
+			}
+		}
+	}
+
+	// TODO: Create a EnvoyFilterContext struct, just like SidecarContext
+	// pre-process the values and convert them into appropriate structs when initializing the push context
+	for _, cp := range filterCRD.ConfigPatches {
+		if cp.Patch == nil ||
+			(cp.ApplyTo != networking.EnvoyFilter_ROUTE_CONFIGURATION &&
+				cp.ApplyTo != networking.EnvoyFilter_VIRTUAL_HOST) {
+			continue
+		}
+
+		if cp.Patch.Operation != networking.EnvoyFilter_Patch_MERGE {
+			continue
+		}
+
+		userChanges, err := buildXDSObjectFromValue(cp.ApplyTo, cp.Patch.Value)
 		if err != nil {
-			return nil, err
+			continue
+		}
+
+		// if applying the merge at routeConfiguration level, then do standard proto merge
+		if cp.ApplyTo == networking.EnvoyFilter_ROUTE_CONFIGURATION {
+			if routeConfigurationMatch(routeConfiguration, nil, pluginParams, cp.Match) {
+				proto.Merge(routeConfiguration, userChanges)
+			}
+		} else {
+			// This is for a specific virtual host. We have to iterate through all the vhosts in a route,
+			// and match the specific virtual host to merge
+			for i := range routeConfiguration.VirtualHosts {
+				if routeConfigurationMatch(routeConfiguration, &routeConfiguration.VirtualHosts[i], pluginParams, cp.Match) {
+					proto.Merge(&routeConfiguration.VirtualHosts[i], userChanges)
+				}
+			}
+		}
+	}
+
+	// Add virtual host if the operation is add, and patch context matches
+	for _, cp := range filterCRD.ConfigPatches {
+		if cp.Patch == nil || cp.ApplyTo != networking.EnvoyFilter_VIRTUAL_HOST {
+			continue
+		}
+
+		if cp.Patch.Operation != networking.EnvoyFilter_Patch_ADD {
+			continue
+		}
+
+		if routeConfigurationMatch(routeConfiguration, nil, pluginParams, cp.Match) {
+			newVirtualHost, err := buildXDSObjectFromValue(cp.ApplyTo, cp.Patch.Value)
+			if err != nil {
+				continue
+			}
+			routeConfiguration.VirtualHosts = append(routeConfiguration.VirtualHosts, *newVirtualHost.(*route.VirtualHost))
+		}
+	}
+
+	if virtualHostsRemoved {
+		trimmedVirtualHosts := make([]route.VirtualHost, 0, len(routeConfiguration.VirtualHosts))
+		for _, virtualHost := range routeConfiguration.VirtualHosts {
+			if virtualHost.Name == "" {
+				continue
+			}
+			trimmedVirtualHosts = append(trimmedVirtualHosts, virtualHost)
+		}
+		routeConfiguration.VirtualHosts = trimmedVirtualHosts
+	}
+
+	return routeConfiguration
+}
+
+func clusterMatch(proxy *model.Proxy, cluster *xdsapi.Cluster,
```

Comment:
do we expect users to only specify one cluster at a time for patching?

Reply:
for adds, yes. user can add only one cluster. and for merges, they specify the pieces of a cluster that needs to be updated.